# Vanessa Williams

## DSC350: Week 4

In [381]:
pip install quandl

In [383]:
import pandas as pd
import quandl
import numpy as np
from numpy.random import seed
from numpy.random import rand
from numpy.random import randint

## Section 1

In [386]:
# Uploading and printing dataframe
from pandas.io.parsers import read_csv
df = read_csv("WHO_first9cols.csv")
print("Dataframe:\n", df)

Dataframe:
                 Country  CountryID  Continent  Adolescent fertility rate (%)  \
0           Afghanistan          1          1                          151.0   
1               Albania          2          2                           27.0   
2               Algeria          3          3                            6.0   
3               Andorra          4          2                            NaN   
4                Angola          5          3                          146.0   
..                  ...        ...        ...                            ...   
197             Vietnam        198          6                           25.0   
198  West Bank and Gaza        199          1                            NaN   
199               Yemen        200          1                           83.0   
200              Zambia        201          3                          161.0   
201            Zimbabwe        202          3                          101.0   

     Adult literacy rate (%

In [388]:
# Printing the number of rows from dataframe
rows = len(df)
print("length of rows: ", rows)

length of rows:  202


In [390]:
# Printing the column headers from dataframe
headers = df.columns
print("column headers:", headers)

column headers: Index(['Country', 'CountryID', 'Continent', 'Adolescent fertility rate (%)',
       'Adult literacy rate (%)',
       'Gross national income per capita (PPP international $)',
       'Net primary school enrolment ratio female (%)',
       'Net primary school enrolment ratio male (%)',
       'Population (in thousands) total'],
      dtype='object')


In [392]:
# Printing the index
print(df.index)

RangeIndex(start=0, stop=202, step=1)


## Section 2

In [395]:
# Printing data tpye of the column Country
print(df['Country'].dtypes)

# Pinting its series shape
print(pd.Series(df.Country).shape)

# Printing its index
print(df.Country.index)

# Printing the values
print(df.Country.values)


object
(202,)
RangeIndex(start=0, stop=202, step=1)
['Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus'
 'Belgium' 'Belize' 'Benin' 'Bermuda' 'Bhutan' 'Bolivia'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'Brunei Darussalam'
 'Bulgaria' 'Burkina Faso' 'Burundi' 'Cambodia' 'Cameroon' 'Canada'
 'Cape Verde' 'Central African Republic' 'Chad' 'Chile' 'China' 'Colombia'
 'Comoros' 'Congo, Dem. Rep.' 'Congo, Rep.' 'Cook Islands' 'Costa Rica'
 "Cote d'Ivoire" 'Croatia' 'Cuba' 'Cyprus' 'Czech Republic' 'Denmark'
 'Djibouti' 'Dominica' 'Dominican Republic' 'Ecuador' 'Egypt'
 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Ethiopia' 'Fiji'
 'Finland' 'France' 'French Polynesia' 'Gabon' 'Gambia' 'Georgia'
 'Germany' 'Ghana' 'Greece' 'Grenada' 'Guatemala' 'Guinea' 'Guinea-Bissau'
 'Guyana' 'Haiti' 'Honduras' 'Hong Kong, China' 'Hungary' 'Iceland'
 'India

## Section 3

In [398]:
import quandl
quandl.ApiConfig.api_key = 's4YsJzqszoSGeJMsHP6_'

In [439]:
import pandas as pd

In [441]:
# Loading data 
data = quandl.get_table("QDL/OPEC")

# Printing the first several rows of the dataframe
print(data.head())

           date  value
None                  
0    2024-01-25  81.98
1    2024-01-24  81.05
2    2024-01-23  81.30
3    2024-01-22  79.70
4    2024-01-19  80.27


In [442]:
# Printing the last several rows of the dataframe
print(data.tail())

           date  value
None                  
5430 2003-01-08  28.86
5431 2003-01-07  29.72
5432 2003-01-06  30.71
5433 2003-01-03  30.83
5434 2003-01-02  30.05


In [445]:
data.dtypes

date     datetime64[ns]
value           float64
dtype: object

In [447]:
# Get the last value (the last row in the dataframe)
last_value = data.iloc[-1]
print("Last value:", last_value)

Last value: date     2003-01-02 00:00:00
value                  30.05
Name: 5434, dtype: object


In [449]:
# Set the "date" column as the index
data.set_index("date", inplace=True)

In [451]:
print(data.head())

            value
date             
2024-01-25  81.98
2024-01-24  81.05
2024-01-23  81.30
2024-01-22  79.70
2024-01-19  80.27


In [453]:
data = data.sort_index()

In [455]:
data.tail()

value
date             
2024-01-19  80.27
2024-01-22  79.70
2024-01-23  81.30
2024-01-24  81.05
2024-01-25  81.98

In [457]:
# Quering the correct date range
data_query = data["2003-01-02":"2023-12-31"]
print(data_query)

            value
date             
2003-01-02  30.05
2003-01-03  30.83
2003-01-06  30.71
2003-01-07  29.72
2003-01-08  28.86
...           ...
2023-12-22  81.23
2023-12-26  80.71
2023-12-27  81.82
2023-12-28  80.31
2023-12-29  78.44

[5417 rows x 1 columns]


In [459]:
# Query with a Boolean, where the number of observations is greater than the mean number of observations
mean_value = data["value"].mean()
print("Mean value:", mean_value)

Mean value: 69.90859797608097


In [461]:
# FIlter rows where the value is greater than the mean
greater_than_mean = data[data["value"] > mean_value]
print(greater_than_mean)

            value
date             
2006-07-13  70.27
2006-07-14  71.71
2006-07-17  71.40
2006-07-18  70.29
2006-08-01  69.97
...           ...
2024-01-19  80.27
2024-01-22  79.70
2024-01-23  81.30
2024-01-24  81.05
2024-01-25  81.98

[2513 rows x 1 columns]


## Part 4

In [470]:
import pandas as pd

# Convert your data into a DataFrame
data = {'col1': [1, 2, 3, 4, 5], 'col2': [10, 20, 30, 40, 50]}
df = pd.DataFrame(data)

# Now you can use .describe()
print(df.describe())

           col1       col2
count  5.000000   5.000000
mean   3.000000  30.000000
std    1.581139  15.811388
min    1.000000  10.000000
25%    2.000000  20.000000
50%    3.000000  30.000000
75%    4.000000  40.000000
max    5.000000  50.000000


In [472]:
import quandl
print(df.describe(),'\n')
print(df.count(),'\n')
print(df.mad(),'\n')
print(df.median(),'\n')
print(df.min(),'\n')
print(df.max(),'\n')
print(df.mode(),'\n')
print(df.std(),'\n')
print(df.var(),'\n')
print(df.skew(),'\n')
print(df.kurt(),'\n')

           col1       col2
count  5.000000   5.000000
mean   3.000000  30.000000
std    1.581139  15.811388
min    1.000000  10.000000
25%    2.000000  20.000000
50%    3.000000  30.000000
75%    4.000000  40.000000
max    5.000000  50.000000 

col1    5
col2    5
dtype: int64 



AttributeError: 'DataFrame' object has no attribute 'mad'

# Help

In [509]:
import numpy as np

# Defining data frame
df = np.array([1, 2, 3, 4, 5])

# Step 1: Calculate the mean of the data
mean = df.mean()

# Step 2: Calculate the absolute deviation from the mean
absolute_deviation = np.abs(df - mean)

# Step 3: Calculate the mean of the absolute deviations (MAD)
mad = absolute_deviation.mean()

print(f'Mean Absolute Deviation: {mad}')

Mean Absolute Deviation: 1.2


## Section 5

In [513]:
import numpy as np
import pandas as pd

# Generate random data
np.random.seed(42)
data = {
    'Weather': np.random.choice(['Hot', 'Cold'], size=100),
    'Food': np.random.choice(['Pizza', 'Burger', 'Salad'], size=100),
    'Price': np.random.uniform(5, 25, size=100), # Random prices between 5 and 25 dollars
    'Number': np.random.randint(1, 10, size=100) # Random integer for number
}
df = pd.DataFrame(data)
print(df.head())

  Weather    Food      Price  Number
0     Hot   Salad  14.130691       5
1    Cold  Burger   9.368809       4
2     Hot  Burger  13.330199       8
3     Hot  Burger  22.665605       8
4     Hot  Burger  11.486900       7


In [519]:
weather_group = df.groupby('Weather')

# Print first row, last row, and mean for each group
for name, group in weather_group:
    print(f"Group: {name}")
    print("First row:\n", group.iloc[0])
    print("Last row:\n", group.iloc[-1])
    print("Mean for Price and Number:\n", group[['Price', 'Number']].mean())
    print("\n")

Group: Cold
First row:
 Weather        Cold
Food         Burger
Price      9.368809
Number            4
Name: 1, dtype: object
Last row:
 Weather        Cold
Food          Pizza
Price      8.518505
Number            3
Name: 98, dtype: object
Mean for Price and Number:
 Price     15.396534
Number     4.482143
dtype: float64


Group: Hot
First row:
 Weather          Hot
Food           Salad
Price      14.130691
Number             5
Name: 0, dtype: object
Last row:
 Weather         Hot
Food         Burger
Price      5.361507
Number            1
Name: 99, dtype: object
Mean for Price and Number:
 Price     15.122423
Number     4.977273
dtype: float64




In [521]:
weather_food_group = df.groupby(['Weather', 'Food'])

# Applying aggregate functions: Mean and Median for 'Number' and 'Price'
agg_result = weather_food_group.agg({
    'Number': ['mean', 'median'],
    'Price': ['mean', 'median']
})

print("\nAggregated results for 'Weather' and 'Food' with mean and median:\n", agg_result)


Aggregated results for 'Weather' and 'Food' with mean and median:
                   Number             Price           
                    mean median       mean     median
Weather Food                                         
Cold    Burger  4.250000    4.0  13.303772  10.892848
        Pizza   3.545455    3.0  14.580279  12.734906
        Salad   5.545455    6.0  17.354295  18.882204
Hot     Burger  4.785714    6.0  13.955085  14.013399
        Pizza   5.384615    6.0  14.884638  13.955663
        Salad   4.823529    5.0  16.265597  15.828959


## Section 6

In [523]:
# Select first three rows
first_three_rows = df.head(3)

In [525]:
# Concatatenate the first three rows with the original data frame
df_concat = pd.concat([df, first_three_rows], ignore_index=True)
print("Concatenated DataFrame:\n", df_concat.tail())

# Concatatenate the first three rows and the last two rows together
last_two_rows = df.tail(2)
df_combined = pd.concat([first_three_rows, last_two_rows], ignore_index=True)
print("\nCombined DataFrame (first three and last two rows):\n", df_combined)

Concatenated DataFrame:
     Weather    Food      Price  Number
98     Cold   Pizza   8.518505       3
99      Hot  Burger   5.361507       1
100     Hot   Salad  14.130691       5
101    Cold  Burger   9.368809       4
102     Hot  Burger  13.330199       8

Combined DataFrame (first three and last two rows):
   Weather    Food      Price  Number
0     Hot   Salad  14.130691       5
1    Cold  Burger   9.368809       4
2     Hot  Burger  13.330199       8
3    Cold   Pizza   8.518505       3
4     Hot  Burger   5.361507       1


In [479]:
# Combining two data frames to create 
pd.concat([dataFrame[:3], dataFrame[3:]])

weather      food     price  number
0    cold      soup  3.745401       8
1     hot  popsicle  9.507143       5
2    cold     steak  7.319939       4
3     hot   seafood  5.986585       8
4    cold      soup  1.560186       8
5     hot     steak  1.559945       3
6    cold  popsicle  0.580836       6

## Section 7

In [485]:
# Reading in the data
dests = pd.read_csv('dest.csv')
tips = pd.read_csv('tips.csv')

# Applying a merge to each data set
new_set = pd.merge(dests, tips, on='EmpNr')
print(new_set) # Printing the new set

   EmpNr       Dest  Amount
0      5  The Hague    10.0
1      9  Rotterdam     5.0


In [487]:
# Applying a join to each data set
new_set2 = dests.join(tips, lsuffix='Dest', rsuffix='tips')
print(new_set2) 

   EmpNrDest       Dest  EmpNrtips  Amount
0          5  The Hague          5    10.0
1          3  Amsterdam          9     5.0
2          9  Rotterdam          7     2.5


## Section 8

In [489]:
df = pd.read_csv('WHO_first9cols.csv')
# Selecting the first three rows of the who data, including the headers for these two columns: Country & Net primary school enrolment ratio male (%)
df = df[['Country', df.columns[-2]]][:2]

# Check for missing values, count the number of NaN values, print any non-missing values, and replace the missing values with a scalar value
print("New df\n", df)
print("Null Values\n", pd.isnull(df))
print("Total Null Values\n", pd.isnull(df).sum())
print("Not Null Values\n", df.notnull())
print("Last Column Doubled\n", 2 * df[df.columns[-1]])
print("Last Column plus NaN\n", df[df.columns[-1]] + np.nan)
print("Zero filled\n", df.fillna(0))

New df
        Country  Net primary school enrolment ratio male (%)
0  Afghanistan                                          NaN
1      Albania                                         94.0
Null Values
    Country  Net primary school enrolment ratio male (%)
0    False                                         True
1    False                                        False
Total Null Values
 Country                                        0
Net primary school enrolment ratio male (%)    1
dtype: int64
Not Null Values
    Country  Net primary school enrolment ratio male (%)
0     True                                        False
1     True                                         True
Last Column Doubled
 0      NaN
1    188.0
Name: Net primary school enrolment ratio male (%), dtype: float64
Last Column plus NaN
 0   NaN
1   NaN
Name: Net primary school enrolment ratio male (%), dtype: float64
Zero filled
        Country  Net primary school enrolment ratio male (%)
0  Afghanistan                

#### Cite:
#### Title: 4.2 Exercise
#### Author: Armando Fandago
#### Date: 10/09/2024
#### Modified By: Isis Lara
#### Description: Follow along code snippets in required reading